### PR TITLE
Solution for Sorting a List of Numbers with Merge Sort Algorithm and Recursion

### LINKED ISSUE #31: Challenge: Sort a List of Numbers with Merge Sort Algorithm and Recursion
## Description

Sort a list of integers and/or floats in the ascending order using the `Merge Sort Algorithm` and `Recursion`.

---

## Behavior

The function should:
- Accepts `a list of integers and/or floats` as an array to be sorted.
- Returns `a list of sorted numbers` in the ascending order.

---

## Strategy for `Merge Sort Algorithm` and `Recursion`

1. **Base case**: An array of 1 or 0 element is already sorted.
2. **Break-down**: Split the array in half.
3. **Recursion**: Sort the left half.
4. **Recursion**: Sort the right half.
5. **Build-up**: Merge the two halves together in a sorted order.

---

### Example 1

    Input:
    function([5, 4, 3, 2, 1, 0, -1, -2])

    Output:
    [-2, -1, 0, 1, 2, 3, 4, 5]

### Example 2

    Input:
    function([1.0, 1, 3, 2.5, 4.5, 5, 7])

    Output:
    [1.0, 1, 2.5, 3, 4.5, 5, 7]

### Example 3

    Input:
    function([2.999, 3.0, 4.1, 4.05, 200, 2000, 90000])

    Output:
    [2.999, 3.0, 4.05, 4.1, 200, 2000, 90000]

### PR DESCRIPTION
**Pull request by Ammmoe**
---
name: Sort a List of Numbers with Merge Sort Algorithm and Recursion
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/merge_sort.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+A module for sorting a list of integers and/or floats in the ascending order using the merge sort algorithm and recursion
+
+Module contents:
+    - merge_sort function
```
Comment (hamid4231) [inline solutions/merge_sort.py]: Hello @Ammmoe, as I am going through your code I noticed the name of your file and function a little bit hard to understand. For someone who doesn't know the merge_sort algorithm, it would be hard to understand your function's behavior. What do you think about coming up with a more descriptive file name?

File: solutions/tests/test_merge_sort.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+import unittest
+
+from ..merge_sort import merge_sort
+
+
+class TestMergeSort(unittest.TestCase):
+    """Test suite for the merge_sort function."""
+
+    # Standard test cases
+    def test_std_1(self):
```
Comment (hamid4231) [inline solutions/tests/test_merge_sort.py]: @Ammmoe, Is std an abbreviation of standard deviation? If so I suggest using the full name instead of abbreviation.

File: solutions/merge_sort.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+Created on 8-Jan-2025
+
+@author: Aung Myin Moe
+"""
+
+
+# ---- define function ----
+def merge_sort(array: list[int, float]) -> list[int, float]:
```
Comment (hamid4231) [inline solutions/merge_sort.py]: @Ammmoe, Here I want to raise the same concern as file naming for your function naming.

**Review by hamid4231 — CHANGES_REQUESTED** [review]
Hello @Ammmoe,
I have reviewed your solution and left some suggestion. Please feel free to check them out and I am open if you need to discuss your ideas.
Best,
Hamid

File: solutions/tests/test_merge_sort.py (hunk 2 above)
Comment (Ammmoe) [reply solutions/tests/test_merge_sort.py]: Hi @hamid4231 yes, by `test_std_1`, I mean `the standard test case no. 1`. 

Would it be OK if I just change it to `test_standard_1` ?

File: solutions/merge_sort.py (hunk 1 above)
Comment (Ammmoe) [reply solutions/merge_sort.py]: I agree with your point @hamid4231 . 

I think maybe the name `sort_number` should be a bit more descriptive.

I will describe `merge_sort` strategy maybe in the comments so reviewers may know the sorting algorithm that I use in this file.

What do you think of this idea?

File: solutions/merge_sort.py (hunk 3 above)
Comment (Ammmoe) [reply solutions/merge_sort.py]: Hi @hamid4231 ! Thanks for pointing this out.

File: solutions/merge_sort.py (hunk 1 above)
Comment (hamid4231) [reply solutions/merge_sort.py]: That sounds great @Ammmoe! I think it will be good to have some descriptions for the viewers.

File: solutions/tests/test_merge_sort.py (hunk 2 above)
Comment (hamid4231) [reply solutions/tests/test_merge_sort.py]: @Ammmoe, It sounds good too, but I think it will be better and more descriptive if you use like: "test_integers, test_floats, test_large_integers, test_large_floats"
What do you think about these namings?

**Review by hamid4231 — CHANGES_REQUESTED** [review]

File: solutions/merge_sort.py (hunk 1 above)
Comment (Ammmoe) [reply solutions/merge_sort.py]: Thanks @hamid4231 ! Now I've changed to a more intuitive file names with easier-to-understand docstring. 

I've mentioned the strategy and algorithm in the code comment.

File: solutions/tests/test_merge_sort.py (hunk 2 above)
Comment (Ammmoe) [reply solutions/tests/test_merge_sort.py]: @hamid4231 I've changed my unittest names to a more intuitive ones like you suggested. Please check and let me know if I need to modify more.

**Review by hamid4231 — APPROVED** [review]
Hello @Ammmoe, your solution looks much better after the changes and now I am going to merge your solution to the main branch. I will delete you challenges branch after the merge.
Thanks for the patience and cooperation.
Best,
Hamid
